File: python/path-crossing.py
MOVES = {
    "N": (1, 0),
    "S": (-1, 0),
    "E": (0, 1),
    "W": (0, -1),
}

class Solution:
    def isPathCrossing(self, path: str) -> bool:
        current = (0, 0)
        visited = set([current])
        for m in path:
            new_position = (current[0] + MOVES[m][0], current[1] + MOVES[m][1])
            if new_position in visited:
                return True
            visited.add(new_position)
            current = new_position
        return False

File: rust/path-crossing.rs
use std::collections::HashMap;
use std::collections::HashSet;

impl Solution {
    pub fn is_path_crossing(path: String) -> bool {
        let DIRECTIONS = HashMap::from([
            ('N', 100000),
            ('S', -100000),
            ('E', 1),
            ('W', -1)
        ]);
        let mut current_position = 0;
        let mut visited = HashSet::new();
        visited.insert(current_position);
        for c in path.chars() {
            current_position += DIRECTIONS[&c];
            if visited.contains(&current_position) {
                return true;
            }
            visited.insert(current_position);
        }
        return false;
    }
}
